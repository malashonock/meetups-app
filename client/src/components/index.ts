export * from './Typography/Typography';
export * from './Button/Button';
export * from './MeetupCard/MeetupCard';
export * from './UserPreview/UserPreview';
export * from './VotesCount/VotesCount';
export * from './IconButton/IconButton';
export * from './DeleteButton/DeleteButton';
export * from './EditButton/EditButton';
export * from './Header/Header';
export * from './Tooltip/Tooltip';
export * from './InputField/InputField';
export * from './TextInput/TextInput';
export * from './TextField/TextField';
export * from './TextArea/TextArea';
export * from './HelperText/HelperText';
export * from './InputLabel/InputLabel';
export * from './Tabs/Tabs';
export * from './NavTabs/NavTabs';
export * from './Tab/Tab';
export * from './TabContent/TabContent';
export * from './TabsManager/TabsManager';
export * from './TabsIndicator/TabsIndicator';
export * from './MeetupStagesTabs/MeetupStagesTabs';
export * from './MeetupStagesTabs/MeetupTabContent/MeetupTabContent';
export * from './MeetupStagesTabs/meetupTabsMapper';
export * from './NewsCard/NewsCard';
export * from './DateTimePicker/DateTimePicker';
export * from './Stepper/Stepper';
export * from './Step/Step';
export * from './StepperProgress/StepperProgress';
export * from './ImageDropbox/ImageDropbox';
export * from './ImagePreview/ImagePreview';
export * from './ImageUploader/ImageUploader';
export * from './AuthToggle/AuthToggle';
export * from './SelectField/SelectField';
export * from './RootStoreProvider/RootStoreProvider';
export * from './LanguageSelect/LanguageSelect';
export * from './ProtectedRoute/ProtectedRoute';
export * from './AvatarGroup/AvatarGroup';
export * from './LoadingSpinner/LoadingSpinner';
export * from './Portal/Portal';
export * from './Modal/Modal';
export * from './ConfirmDialog/ConfirmDialog';
export * from './ConfirmDialogProvider/ConfirmDialogProvider';
export * from './Toast/Toast';
export * from './AlertStack/AlertStack';
