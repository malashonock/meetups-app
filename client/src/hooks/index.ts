export * from './useAuthStore';
export * from './useUserStore';
export * from './useUser';
export * from './useMeetupStore';
export * from './useMeetup';
export * from './useNewsStore';
export * from './useNewsArticle';
export * from './useUiStore';
export * from './useLocale';
export * from './useTouchOnLocaleChanged';
export * from './useResizeEffect';
export * from './useConfirmDialog';
