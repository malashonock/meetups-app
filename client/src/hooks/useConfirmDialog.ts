import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
