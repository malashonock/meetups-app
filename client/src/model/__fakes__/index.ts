export * from './user.fake';
export * from './meetup.fake';
export * from './file.fake';
