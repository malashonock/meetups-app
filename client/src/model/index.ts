export * from './meetup.model';
export * from './news.model';
export * from './user.model';
export * from './login.model';
