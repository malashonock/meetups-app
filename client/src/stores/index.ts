export * from './root.store';
export * from './auth.store';
export * from './user.store';
export * from './ui.store';
export * from './meetup.store';
export * from './news.store';
