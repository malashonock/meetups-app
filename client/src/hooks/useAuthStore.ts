import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
