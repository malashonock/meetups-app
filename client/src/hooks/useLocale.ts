import { Locale } from 'stores';
import { useUiStore } from 'hooks';

export type UseLocaleResult = [
  locale: Locale,
  setLocale: (locale: Locale) => void,
];

export const useLocale = (): UseLocaleResult => {
  const uiStore = useUiStore();
  const { locale, setLocale } = uiStore;
  return [locale, setLocale.bind(uiStore)];
};
