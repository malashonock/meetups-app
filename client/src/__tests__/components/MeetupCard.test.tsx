import { PropsWithChildren } from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

import { ConfirmDialogProvider, MeetupCard } from 'components';
import { mockFullUser, mockMeetup, mockTopic } from 'model/__fakes__';
import { useAuthStore, useLocale } from 'hooks';
import { Locale, Meetup, RootStore } from 'stores';

// Mock hooks
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useAuthStore: jest.fn(),
    useLocale: jest.fn(),
  };
});
const mockUseAuthStore = useAuthStore as jest.MockedFunction<
  typeof useAuthStore
>;
const mockUseLocale = useLocale as jest.MockedFunction<typeof useLocale>;

const MockLoginRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <ConfirmDialogProvider>
    <MemoryRouter initialEntries={['/meetups']}>
      <Routes>
        <Route path="/meetups">
          <Route index element={children} />
          <Route path=":id/edit" element={<h1>Edit meetup</h1>} />
        </Route>
      </Routes>
    </MemoryRouter>
  </ConfirmDialogProvider>
);

beforeEach(() => {
  const { authStore } = new RootStore();
  authStore.loggedUser = mockFullUser;
  mockUseAuthStore.mockReturnValue(authStore);

  mockUseLocale.mockReturnValue([Locale.RU, jest.fn()]);
});

afterEach(() => {
  jest.resetAllMocks();
});

describe('MeetupCard', () => {
  describe('before topic approval', () => {
    it('should match snapshot', () => {
      const { asFragment } = render(<MeetupCard meetup={mockTopic} />, {
        wrapper: MockLoginRouter,
      });

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render author, topic & decription, votes count', () => {
      render(<MeetupCard meetup={mockTopic} />, { wrapper: MockLoginRouter });

      const author = screen.getByText('John Doe');
      expect(author).toBeInTheDocument();

      const topic = screen.getByText('Test meetup topic');
      expect(topic).toBeInTheDocument();

      const description = screen.getByText('Test meetup description');
      expect(description).toBeInTheDocument();

      const votesCount = screen.getByText('votesCount.text');
      expect(votesCount).toBeInTheDocument();
    });
  });

  describe('after topic approval', () => {
    it('should match snapshot', () => {
      const { asFragment } = render(<MeetupCard meetup={mockMeetup} />, {
        wrapper: MockLoginRouter,
      });

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render date/time & location, topic & decription, and author', () => {
      render(<MeetupCard meetup={mockMeetup} />, { wrapper: MockLoginRouter });

      const date = screen.getByText('15 марта', { exact: false });
      expect(date).toBeInTheDocument();

      const time = screen.getByText('14:00');
      expect(time).toBeInTheDocument();

      const location = screen.getByText('room 123');
      expect(location).toBeInTheDocument();

      const topic = screen.getByText('Test meetup topic');
      expect(topic).toBeInTheDocument();

      const description = screen.getByText('Test meetup description');
      expect(description).toBeInTheDocument();

      const author = screen.getByText('John Doe');
      expect(author).toBeInTheDocument();
    });
  });

  describe('edit & delete buttons', () => {
    it('should not be rendered unless user is logged in', () => {
      const { authStore } = new RootStore();
      authStore.loggedUser = null;
      mockUseAuthStore.mockReturnValue(authStore);

      render(<MeetupCard meetup={mockMeetup} />, { wrapper: MockLoginRouter });

      const deleteButton = screen.queryByTestId('delete-button');
      expect(deleteButton).toBeNull();

      const editButton = screen.queryByTestId('edit-button');
      expect(editButton).toBeNull();
    });

    it('before topic approval, only delete button should be rendered', () => {
      render(<MeetupCard meetup={mockTopic} />, { wrapper: MockLoginRouter });

      const deleteButton = screen.getByTestId('delete-button');
      expect(deleteButton).toBeInTheDocument();

      const editButton = screen.queryByTestId('edit-button');
      expect(editButton).toBeNull();
    });

    it('after topic approval, both delete and edit buttons should be rendered', () => {
      render(<MeetupCard meetup={mockMeetup} />, { wrapper: MockLoginRouter });

      const deleteButton = screen.getByTestId('delete-button');
      expect(deleteButton).toBeInTheDocument();

      const editButton = screen.getByTestId('edit-button');
      expect(editButton).toBeInTheDocument();
    });
  });

  it('on delete button click, should call meetup.delete() method', async () => {
    const spiedOnDelete = jest.spyOn(Meetup.prototype, 'delete');

    render(<MeetupCard meetup={mockMeetup} />, { wrapper: MockLoginRouter });

    userEvent.click(screen.getByTestId('delete-button'));
    userEvent.click(screen.getByTestId('confirm-button'));

    await waitFor(() => {
      expect(spiedOnDelete).toHaveBeenCalledTimes(1);
    });
  });

  it('on edit button click, should navigate to Edit meetup page', async () => {
    render(<MeetupCard meetup={mockMeetup} />, { wrapper: MockLoginRouter });

    const editButton = screen.getByTestId('edit-button');
    userEvent.click(editButton);

    const editMeetupPage = await screen.findByText('Edit meetup');
    expect(editMeetupPage).toBeInTheDocument();
  });
});
