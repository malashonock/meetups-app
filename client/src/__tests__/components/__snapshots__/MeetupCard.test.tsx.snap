// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetupCard after topic approval should match snapshot 1`] = `
<DocumentFragment>
  <article
    class="card upcoming"
    data-testid="meetup-card"
  >
    <header
      class="header"
    >
      <ul
        class="appointment"
      >
        <li
          class="appointmentItem"
        >
          <span
            class="typography span date noWrap"
          >
            f, 15 марта
          </span>
        </li>
        <li
          class="appointmentItem"
        >
          <span
            class="typography span time noWrap"
          >
            14:00
          </span>
        </li>
        <li
          class="appointmentItem"
        >
          <span
            class="typography span location noWrap"
          >
            room 123
          </span>
        </li>
      </ul>
      <div
        class="controls"
      >
        <button
          class="button"
          data-testid="delete-button"
        >
          <svg>
            delete.svg
          </svg>
        </button>
        <button
          class="button"
          data-testid="edit-button"
        >
          <svg>
            edit.svg
          </svg>
        </button>
      </div>
    </header>
    <div
      class="body"
    >
      <h2
        class="typography h2 subject"
      >
        Test meetup topic
      </h2>
      <p
        class="typography p excerpt"
      >
        Test meetup description
      </p>
    </div>
    <footer
      class="footer"
    >
      <div
        class="user card"
        data-testid="user-preview"
      >
        <div
          class="avatar"
        >
          <span
            class="typography span initials"
          >
            JD
          </span>
        </div>
        <span
          class="typography span name noWrap"
        >
          John Doe
        </span>
      </div>
    </footer>
  </article>
</DocumentFragment>
`;

exports[`MeetupCard before topic approval should match snapshot 1`] = `
<DocumentFragment>
  <article
    class="card topic"
    data-testid="meetup-card"
  >
    <header
      class="header"
    >
      <div
        class="user card"
        data-testid="user-preview"
      >
        <div
          class="avatar"
        >
          <span
            class="typography span initials"
          >
            JD
          </span>
        </div>
        <span
          class="typography span name noWrap"
        >
          John Doe
        </span>
      </div>
      <div
        class="controls"
      >
        <button
          class="button"
          data-testid="delete-button"
        >
          <svg>
            delete.svg
          </svg>
        </button>
      </div>
    </header>
    <div
      class="body"
    >
      <h2
        class="typography h2 subject"
      >
        Test meetup topic
      </h2>
      <p
        class="typography p excerpt"
      >
        Test meetup description
      </p>
    </div>
    <footer
      class="footer"
    >
      <div
        class="wrapper"
      >
        <svg
          class="icon"
        >
          profile.svg
        </svg>
        <p
          class="typography p text"
        >
          votesCount.text
        </p>
      </div>
    </footer>
  </article>
</DocumentFragment>
`;
